File: bindings/python/src/graphqlite/algorithms/traversal.py
"""Graph traversal algorithms mixin."""

from ..graph._base import BaseMixin
from ._parsing import extract_algo_array, parse_traversal_result


class TraversalMixin(BaseMixin):
    """Mixin providing graph traversal algorithm methods."""

    def bfs(
        self,
        start_id: str,
        max_depth: int = -1
    ) -> list[dict]:
        """
        Perform breadth-first search traversal from a starting node.

        BFS explores nodes level by level, visiting all neighbors at depth d
        before any nodes at depth d+1. Useful for finding shortest paths in
        unweighted graphs.

        Args:
            start_id: Starting node's id property value
            max_depth: Maximum depth to traverse (-1 for unlimited)

        Returns:
            List of dicts with 'user_id', 'depth', 'order' sorted by traversal order
        """
        if max_depth < 0:
            query = f"RETURN bfs('{self._escape(start_id)}')"
        else:
            query = f"RETURN bfs('{self._escape(start_id)}', {max_depth})"

        result = self._conn.cypher(query)
        rows = extract_algo_array(result.to_list())

        nodes = []
        for row in rows:
            parsed = parse_traversal_result(row)
            if parsed is not None:
                nodes.append(parsed)

        return nodes

    # Alias
    breadth_first_search = bfs

    def dfs(
        self,
        start_id: str,
        max_depth: int = -1
    ) -> list[dict]:
        """
        Perform depth-first search traversal from a starting node.

        DFS explores as far as possible along each branch before backtracking.
        Useful for topological sorting, detecting cycles, and exploring paths.

        Args:
            start_id: Starting node's id property value
            max_depth: Maximum depth to traverse (-1 for unlimited)

        Returns:
            List of dicts with 'user_id', 'depth', 'order' sorted by traversal order
        """
        if max_depth < 0:
            query = f"RETURN dfs('{self._escape(start_id)}')"
        else:
            query = f"RETURN dfs('{self._escape(start_id)}', {max_depth})"

        result = self._conn.cypher(query)
        rows = extract_algo_array(result.to_list())

        nodes = []
        for row in rows:
            parsed = parse_traversal_result(row)
            if parsed is not None:
                nodes.append(parsed)

        return nodes

    # Alias
    depth_first_search = dfs

File: bindings/rust/src/algorithms/traversal.rs
//! Graph traversal algorithm implementations.

use super::parsing::{extract_int, extract_string};
use super::TraversalResult;
use crate::graph::Graph;
use crate::utils::escape_string;
use crate::Result;

impl Graph {
    /// Perform breadth-first search traversal from a starting node.
    ///
    /// # Arguments
    ///
    /// * `start_id` - Starting node's id
    /// * `max_depth` - Maximum depth to traverse (None for unlimited)
    pub fn bfs(&self, start_id: &str, max_depth: Option<i32>) -> Result<Vec<TraversalResult>> {
        let esc_start = escape_string(start_id);

        let query = match max_depth {
            Some(depth) => format!("RETURN bfs('{}', {})", esc_start, depth),
            None => format!("RETURN bfs('{}')", esc_start),
        };

        let result = self.connection().cypher(&query)?;

        let mut nodes = Vec::new();
        for row in result.iter() {
            if let Some(user_id) = extract_string(row, "user_id") {
                nodes.push(TraversalResult {
                    user_id,
                    depth: extract_int(row, "depth"),
                    order: extract_int(row, "order"),
                });
            }
        }
        Ok(nodes)
    }

    /// Perform depth-first search traversal from a starting node.
    ///
    /// # Arguments
    ///
    /// * `start_id` - Starting node's id
    /// * `max_depth` - Maximum depth to traverse (None for unlimited)
    pub fn dfs(&self, start_id: &str, max_depth: Option<i32>) -> Result<Vec<TraversalResult>> {
        let esc_start = escape_string(start_id);

        let query = match max_depth {
            Some(depth) => format!("RETURN dfs('{}', {})", esc_start, depth),
            None => format!("RETURN dfs('{}')", esc_start),
        };

        let result = self.connection().cypher(&query)?;

        let mut nodes = Vec::new();
        for row in result.iter() {
            if let Some(user_id) = extract_string(row, "user_id") {
                nodes.push(TraversalResult {
                    user_id,
                    depth: extract_int(row, "depth"),
                    order: extract_int(row, "order"),
                });
            }
        }
        Ok(nodes)
    }
}
